File: packages/core/src/types/market.ts
// Candle (OHLCV) data point
export interface Candle {
  timestamp: string;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

// Order book level
export interface OrderBookLevel {
  price: number;
  size: number;
}

// Order book snapshot
export interface OrderBook {
  bids: OrderBookLevel[];
  asks: OrderBookLevel[];
  timestamp: string;
}

// Single token market data
export interface TokenMarketData {
  symbol: string;
  price: number;
  change24h: number;
  volume24h: number;
  high24h: number;
  low24h: number;
  candles1h: Candle[];
  candles15m: Candle[];
  orderBook?: OrderBook;
  timestamp: string;
}

// Full market snapshot sent to Python each cycle
export interface MarketSnapshot {
  timestamp: string;
  tokens: Record<string, TokenMarketData>;
  portfolio: PortfolioState;
  activeStrategies: string[];
  strategyWeights?: Record<string, number>;
  cycleId: string;
}

export interface PortfolioPosition {
  token: string;
  amount: number;
  avgCostUsd: number;
  currentValueUsd: number;
}

export interface PortfolioState {
  walletAddress: string;
  nativeBalance: number;       // ETH
  stableBalance: number;       // USDC
  positions: PortfolioPosition[];
  totalValueUsd: number;
  dailyPnlUsd: number;
  timestamp: string;
}

File: python/models/market.py
from __future__ import annotations
from typing import Any
from pydantic import BaseModel


class Candle(BaseModel):
    timestamp: str
    open: float
    high: float
    low: float
    close: float
    volume: float


class OrderBookLevel(BaseModel):
    price: float
    size: float


class OrderBook(BaseModel):
    bids: list[OrderBookLevel]
    asks: list[OrderBookLevel]
    timestamp: str


class TokenMarketData(BaseModel):
    symbol: str
    price: float
    change24h: float
    volume24h: float
    high24h: float
    low24h: float
    candles1h: list[Candle] = []
    candles15m: list[Candle] = []
    orderBook: OrderBook | None = None
    timestamp: str


class PortfolioPosition(BaseModel):
    token: str
    amount: float
    avgCostUsd: float
    currentValueUsd: float


class PortfolioState(BaseModel):
    walletAddress: str
    nativeBalance: float
    stableBalance: float
    positions: list[PortfolioPosition] = []
    totalValueUsd: float
    dailyPnlUsd: float
    timestamp: str


class MarketSnapshot(BaseModel):
    timestamp: str
    tokens: dict[str, TokenMarketData]
    portfolio: PortfolioState
    activeStrategies: list[str]
    strategyWeights: dict[str, float] | None = None
    cycleId: str
